### PR TITLE
Implement Line Shape, Scale, and Interpolator

### PR DESCRIPTION
**Pull request by cedrickcooke**
Sample has been updated to implement a subset of this example: https://bl.ocks.org/mbostock/0533f44f2cfabecc5e3a. Ignore the big warning at the top of that page, this specific sample hasn't yet been migrated to observablehq yet and is still directly linked in the d3 docs.

JS/Web Canvas:
sine-wave

Android:
device-2021-12-01-143410

Direct link to sample code: https://github.com/JuulLabs/krayon/blob/cedrickc/shape-data/sample/src/commonMain/kotlin/SineWave.kt

### CODE REVIEW

**Comment by cedrickcooke** [comment]
Note to reviewers: CI is failing because I don't want to re-apply fixes in #35, things should get fixed once I merge that.

**Review by twyatt — COMMENTED** [review]
Pass 1/x

> 34 / 51 files viewed

File: scale/src/commonMain/kotlin/LinearScale.kt
Comment (cedrickcooke) [reply scale/src/commonMain/kotlin/LinearScale.kt:1]: Nope, my bad. 😅 . This was a side effect of me drifting slightly from D3's implementation, but having preemptively made files that were a closer 1:1 match. We'll see if I need to adopt their patterns more closely when I go to implement log- and power- scales, but for getting linear- working to unblock our immediate use cases this seemed a better approach.

File: scale/src/commonMain/kotlin/Scale.kt
Comment (cedrickcooke) [reply scale/src/commonMain/kotlin/Scale.kt:6]: I'll fix this. I'm not even sure exactly what I was originally trying to write there.

**Review by davidtaylor-juul — APPROVED** [review]
LGTM

File: sample/src/commonMain/kotlin/SineWave.kt
Comment (twyatt) [inline sample/src/commonMain/kotlin/SineWave.kt:39]: nit: IMHO, this becomes much more readable with some newlines.

```suggestion
    drawPath(
        path = buildPath(line.render(data.filterNotNull())),
        paint = buildPaint(Paint.Stroke(darkSlateBlue, 0.5f, dash = Paint.Stroke.Dash.Pattern(5f, 5.5f)))
    )
    drawPath(
        path = buildPath(line.render(data)),
        paint = buildPaint(Paint.Stroke(steelBlue, 1f))
    )
```

File: sample/src/commonMain/kotlin/SineWave.kt
Comment (twyatt) [reply sample/src/commonMain/kotlin/SineWave.kt:27]: Oh, nevermind, I think I see now.
The 2π just gives a nice devision. Whereas you'll have even distribution of points through the curve, starting and ending _exactly_ at the beginning of a curve.

File: sample/src/commonMain/kotlin/SineWave.kt
Comment (twyatt) [inline sample/src/commonMain/kotlin/SineWave.kt:50]: I don't quite follow why in this example we have to scale twice (here and lines 35-36) whereas in the linked D3 example they only scaled once.

**Review by twyatt — APPROVED** [review]

File: interpolate/src/commonMain/kotlin/Time.kt
Comment (cedrickcooke) [reply interpolate/src/commonMain/kotlin/Time.kt:19]: > I assume your comfortable with Krayon then being opinionated on the matter?

Yup. Since Krayon is only worried about _display_, I think we can get away with ignoring most of the nuance. If an axis happens to cross a daylight-saving time change, I would still expect to see time increasing linearly on the axis and not repeat or skip a value. Measurements that happen to advance monotonically in real-world-time would just not be monotonic along the X axis. Excuse the low quality hand paint, but something like this:

art

> the user may not expect the behavior we've chosen

This is true. But we either make a decision (like I did here), or don't provide the functionality at all (like KotlinX DateTime). All of our interfaces are public, so if a user really needed another option, they can implement it. For users who don't care, the library provides a sane/reasonable default.

Comment (twyatt) [reply interpolate/src/commonMain/kotlin/Time.kt:19]: Such beautiful art. Where can I buy the NFT? 😝

File: sample/src/commonMain/kotlin/SineWave.kt
Comment (cedrickcooke) [reply sample/src/commonMain/kotlin/SineWave.kt:50]: JS is hiding the call to scale behind function pointers. The equivalent line (only gonna look at x, since y is just the same thing with different function names) is `.attr("cx", line.x())`. That call to `line.x()` just retrieves the function passed to `line.x(` earlier, `function(d) { return x(d.x); }`. Because `x` is a raw function in JS, but an interface in this Kotlin, the equivalent becomes `.each { (d) -> x.scale(d.x) }`. And then I've collapsed a few `.each` calls into a single call, because I think it looks nicer.

Comment (cedrickcooke) [reply sample/src/commonMain/kotlin/SineWave.kt:50]: If that explanation doesn't make sense over text I'm happy to hop on a call. I was really struggling to figure out a nice way to write that.

Comment (twyatt) [reply sample/src/commonMain/kotlin/SineWave.kt:50]: Your explanation was great. I follow now. Thank you!

**Pull request merged**
